### PR TITLE
Support flash attention on 2d attention mask for gpt2 left padding.

### PR DESCRIPTION
**Pull request by zhanghuanrong**

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/attention.cc [hunk 1]
```diff
@@ -89,13 +89,14 @@ Status Attention<T>::ComputeInternal(OpKernelContext* context) const {
   // Check whether we can use fused kernel
   int sm = device_prop.major * 10 + device_prop.minor;
   bool is_mask_1d_seq_len = parameters.mask_type == AttentionMaskType::MASK_1D_KEY_SEQ_LEN;
+  bool is_2d_mask = nullptr != mask_index && mask_index->Shape().NumDimensions() == 2;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention.cc:92]: use parameters.mask_type==AttentionMaskType::MASK_2D_KEY_PADDING instead.

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu [hunk 2]
```diff
@@ -361,6 +362,46 @@ void LaunchTrtSequenceOffset(int* trt_mha_padding_offset,
...
+                           const int sequence_length) {
+
+    typedef cub::BlockReduce<int, kMAX_THREADS_PER_BLOCK> BlockReduce;
+    __shared__ typename BlockReduce::TempStorage temp_storage;
+
+    const int batch_id = blockIdx.x;
+    const int* batch_mask = attention_masks + (batch_id * sequence_length);
+    int first_non_padding = sequence_length;
+    for (int i = threadIdx.x; i < sequence_length; i += blockDim.x) {
+      if (batch_mask[i]) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:379]: Need to handle the following mask
[[1,1, 1, 0, 0, 0],  // right side padding
 [0, 0, 0, 1, 1, 1], // left side padding
 [1, 1, 1, 1, 1, 1]] // no padding.
 
 Need a way to tell that mask is not supported by fused kernel like
[[1, 1, 0, 0, 1, 1]]

Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/bert_padding.cu]: support both left and right, reasonable requirements.

File: onnxruntime/contrib_ops/cuda/bert/attention.cc [hunk 3]
```diff
@@ -89,13 +89,14 @@ Status Attention<T>::ComputeInternal(OpKernelContext* context) const {
   // Check whether we can use fused kernel
   int sm = device_prop.major * 10 + device_prop.minor;
   bool is_mask_1d_seq_len = parameters.mask_type == AttentionMaskType::MASK_1D_KEY_SEQ_LEN;
+  bool is_mask_2d_key_padding = parameters.mask_type == AttentionMaskType::MASK_2D_KEY_PADDING;
 
   if (is_unidirectional_) {  // GPT
     // Fused kernels requires left side padding (The mask shall be sequence lengths or no mask)
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention.cc:95]: Update comment: Fused kernels requires one-side padding: mask can be non (no padding), 1D sequence lengths (right side padding) or 2d mask (either left or right-side padding).

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu [hunk 4]
```diff
@@ -361,6 +362,57 @@ void LaunchTrtSequenceOffset(int* trt_mha_padding_offset,
...
+          last_non_padding = i;
+        }
+      }
+    }
+
+    int left_most = BlockReduce(temp_storage).Reduce(first_non_padding, cub::Min(), blockDim.x);
+    int right_most = BlockReduce(temp_storage).Reduce(last_non_padding, cub::Max(), blockDim.x);
+
+    if (threadIdx.x == 0) {
+      if (left_padding == 0) { // treat default as right padding
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:393]: left_most == 0?

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu [hunk 5]
```diff
@@ -361,6 +362,57 @@ void LaunchTrtSequenceOffset(int* trt_mha_padding_offset,
...
+        trt_mha_padding_offset[2 * batch_id] = (batch_id * sequence_length);
+        trt_mha_padding_offset[2 * batch_id + 1] = (batch_id * sequence_length) + right_most + 1;
+      } else {
+        trt_mha_padding_offset[2 * batch_id] = (batch_id * sequence_length) + left_most;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:397]: trt_mha_padding_offset[2 * batch_id] shall always be batch_id * sequence_length no matter left side or right side padding.

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu [hunk 6]
```diff
@@ -361,6 +362,57 @@ void LaunchTrtSequenceOffset(int* trt_mha_padding_offset,
   }
 }
 
+__global__ void __launch_bounds__(kMAX_THREADS_PER_BLOCK)
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:365]: Add a comment that this kernel assumes left side or right-side padding. It does not support mask with more than 2 segments of ones and zeros like [1, 1, 0, 0, 1, 1] or [0, 0, 1, 1, 0, 0]

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu [hunk 7]
```diff
@@ -361,6 +362,55 @@ void LaunchTrtSequenceOffset(int* trt_mha_padding_offset,
...
+    }
+
+    int leading_count = BlockReduce(temp_storage).Reduce(biggest_position, cub::Max(), blockDim.x);
+
+    if (threadIdx.x == 0) {
+      if (!counting_on_zero) { // default right side padding
+        trt_mha_padding_offset[2 * batch_id] = (batch_id * sequence_length);
+        trt_mha_padding_offset[2 * batch_id + 1] = (batch_id * sequence_length) + leading_count + 1;
+      } else { // left side padding
+        trt_mha_padding_offset[2 * batch_id] = (batch_id * sequence_length) + leading_count + 1;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:395]: Could you add a unit test for left side padding and test it enable & disable fused causal?

File: onnxruntime/contrib_ops/cuda/bert/attention.cc (hunk 3 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention.cc:95]: done

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu (hunk 6 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:368]: Let's use different checkin for the test cases.

File: onnxruntime/contrib_ops/cuda/bert/bert_padding.cu (hunk 7 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/bert_padding.cu:395]: Let's use different checkin for the test cases.

**Review by tianleiwu — APPROVED** [review]
